// @microsoft/generator-sharepoint/lib/generators/onPrem19Webpart/templates/knockout/{componentName}ViewModel.ts
import * as ko from 'knockout';
import styles from './<%= componentName %>.module.scss';
import { I<%= componentClassName %>Props } from './<%= componentClassName %>';

export interface I<%= componentName %>BindingContext extends I<%= componentClassName %>Props {
  shouter: KnockoutSubscribable<{}>;
}

export default class <%= componentName %>ViewModel {
  public description: KnockoutObservable<string> = ko.observable('');

  public <%= componentNameCamelCase %>Class: string = styles.<%= componentNameCamelCase %>;
  public containerClass: string = styles.container;
  public rowClass: string = styles.row;
  public columnClass: string = styles.column;
  public titleClass: string = styles.title;
  public subTitleClass: string = styles.subTitle;
  public descriptionClass: string = styles.description;
  public buttonClass: string = styles.button;
  public labelClass: string = styles.label;

  constructor(bindings: I<%= componentName %>BindingContext) {
    this.description(bindings.description);

    // When web part description is updated, change this view model's description.
    bindings.shouter.subscribe((value: string) => {
      this.description(value);
    }, this, 'description');
  }
}
