// @microsoft/generator-sharepoint/lib/generators/onPrem19FieldCustomizer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Utils = require("../../common/utilities");
const BaseExtension = require("../extension/BaseExtensionGenerator");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    BaseExtension.defineOptions(generator, 'Field Customizer');
}
exports.defineOptions = defineOptions;
class FieldCustomizerGenerator extends BaseExtension.BaseExtensionGenerator {
    // tslint:disable-next-line:no-any
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Field Customizer';
        this.codeName = 'FieldCustomizer';
        this.allowedFrameworks = ['react'];
        defineOptions(this);
    }
    initializing() { }
    prompting() {
        this.ensureCorrectFolder();
        if (!this.shouldExecute()) {
            return Promise.resolve();
        }
        return super.prompting();
    }
    configuring() {
        if (this.shouldExecute()) {
            super.configuring();
            this.ensureDependencyGroup('fieldCustomizer');
            this.context.fieldId = Utils.generateGuid();
            this.serveJson.addConfiguration(this.context.componentNameCamelCase, {
                pageUrl: 'https://contoso.sharepoint.com/sites/mySite/SitePages/myPage.aspx',
                fieldCustomizers: {
                    'InternalFieldName': {
                        id: this.context.componentId,
                        properties: {
                            sampleText: 'Value'
                        }
                    }
                }
            });
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing();
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() { }
    includeClientSideInstances() {
        return false;
    }
    shouldExecute() {
        return this.config.get('environment') === 'onprem19' &&
            this.config.get('componentType') === 'extension' &&
            this.config.get('extensionType') === 'FieldCustomizer';
    }
}
exports.FieldCustomizerGenerator = FieldCustomizerGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = FieldCustomizerGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
