// @microsoft/generator-sharepoint/lib/generators/onPremWebpart/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Utils = require("../../common/utilities");
const BaseComponent = require("../component/BaseComponentGenerator");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    BaseComponent.defineOptions(generator, 'WebPart');
}
exports.defineOptions = defineOptions;
class WebpartGenerator extends BaseComponent.BaseComponentGenerator {
    // tslint:disable-next-line:no-any
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Web part (on-prem)';
        this.codeName = 'WebPart';
        this.allowedFrameworks = ['knockout', 'react'];
        this.folderName = 'webparts';
        defineOptions(this);
    }
    initializing() { }
    prompting() {
        this.ensureCorrectFolder();
        if (!this.shouldExecute()) {
            return Promise.resolve();
        }
        return super.prompting();
    }
    configuring() {
        if (this.shouldExecute()) {
            this.ensureDependencyGroup(this.context.framework);
            this.ensureDependencyGroup('webpart');
            if (!this.configJson.data) {
                throw new Error(`Missing required file: "config/config.json"`);
            }
            try {
                /* tslint:disable:max-line-length */
                this.configJson.addEntry({
                    components: [{
                            entrypoint: `./lib/webparts/${this.context.componentNameCamelCase}/${this.context.componentName}WebPart.js`,
                            manifest: `./src/webparts/${this.context.componentNameCamelCase}/${this.context.componentName}WebPart.manifest.json`
                        }]
                }, this.context.componentClassNameKebabCase);
                /* tslint:enable:max-line-length */
            }
            catch (error) {
                // @todo move this error to occur during prompting
                this.log(`Cannot create two ${this.friendlyName}s with the same name.`);
                throw error;
            }
            this.configJson.addLocalizedResource(this.context.componentStrings, `${this.folderName}/${this.context.componentNameCamelCase}/loc/{locale}.js`);
            this.serveJson.setUpForWebParts();
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing(true);
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() { }
    shouldExecute() {
        return this.config.get('environment') === 'onprem' &&
            this.config.get('componentType') === 'webpart';
    }
}
exports.WebpartGenerator = WebpartGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = WebpartGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
