// @microsoft/generator-sharepoint/lib/generators/library/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Utils = require("../../common/utilities");
const BaseComponent = require("../component/BaseComponentGenerator");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    BaseComponent.defineOptions(generator, 'Library');
}
exports.defineOptions = defineOptions;
class LibraryGenerator extends BaseComponent.BaseComponentGenerator {
    // tslint:disable-next-line:no-any
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Library';
        this.codeName = 'Library';
        this.allowedFrameworks = [];
        this.folderName = 'libraries';
        defineOptions(this);
    }
    initializing() { }
    prompting() {
        this.ensureCorrectFolder();
        if (!this.shouldExecute()) {
            return Promise.resolve();
        }
        return super.prompting();
    }
    configuring() {
        if (this.shouldExecute()) {
            super.configuring();
            this.ensureDependencyGroup('library');
            this.packageJson.data.private = false;
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing(true);
            this.copyTemplate(this.templatePath('./index/'), this.destinationPath('./src'));
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() { }
    shouldExecute() {
        return this.config.get('environment') === 'spo' &&
            this.config.get('componentType') === 'library';
    }
}
exports.LibraryGenerator = LibraryGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = LibraryGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
