// @microsoft/generator-sharepoint/lib/common/YeomanConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable-next-line:typedef
const _strings = {
    libraryName: 'libraryName',
    framework: 'framework',
    version: 'version',
    libraryId: 'libraryId',
    environment: 'environment',
    packageManager: 'packageManager'
};
/**
 * This class wraps an untyped property bag with types for
 * the generator.
 */
class YeomanConfiguration {
    static setStore(yeomanConfiguration) {
        YeomanConfiguration._store = yeomanConfiguration;
    }
    static get libraryName() {
        return YeomanConfiguration._store.get(_strings.libraryName);
    }
    static set libraryName(value) {
        YeomanConfiguration._store.set(_strings.libraryName, value);
    }
    static get version() {
        return YeomanConfiguration._store.get(_strings.version);
    }
    static set version(value) {
        YeomanConfiguration._store.set(_strings.version, value);
    }
    static get libraryId() {
        return YeomanConfiguration._store.get(_strings.libraryId);
    }
    static set libraryId(value) {
        YeomanConfiguration._store.set(_strings.libraryId, value);
    }
    static get environment() {
        return YeomanConfiguration._store.get(_strings.environment);
    }
    static set environment(value) {
        YeomanConfiguration._store.set(_strings.environment, value);
    }
    static get packageManager() {
        return YeomanConfiguration._store.get(_strings.packageManager);
    }
    static set packageManager(value) {
        YeomanConfiguration._store.set(_strings.packageManager, value);
    }
}
exports.YeomanConfiguration = YeomanConfiguration;
