// @microsoft/generator-sharepoint/lib/generators/searchQueryModifier/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Utils = require("../../common/utilities");
const BaseExtension = require("../extension/BaseExtensionGenerator");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    BaseExtension.defineOptions(generator, 'Search Query Modifier');
}
exports.defineOptions = defineOptions;
class SearchQueryModifierGenerator extends BaseExtension.BaseExtensionGenerator {
    // tslint:disable-next-line:no-any
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Search Query Modifier';
        this.codeName = 'SearchQueryModifier';
        this.allowedFrameworks = [];
        defineOptions(this);
    }
    initializing() { }
    prompting() {
        this.ensureCorrectFolder();
        if (!this.shouldExecute()) {
            return Promise.resolve();
        }
        return super.prompting();
    }
    configuring() {
        if (this.shouldExecute()) {
            super.configuring();
            this.ensureDependencyGroup('searchQueryModifier');
            this.serveJson.addConfiguration(this.context.componentNameCamelCase, {
                pageUrl: 'https://contoso.sharepoint.com/sites/mySite/SitePages/myPage.aspx',
                customActions: {
                    [this.context.componentId]: {
                        location: 'ClientSideExtension.SearchQueryModifier',
                        properties: {
                            testMessage: 'Test message'
                        }
                    }
                }
            });
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing();
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() { }
    includeClientSideInstances() {
        return true;
    }
    shouldExecute() {
        return this.config.get('environment') === 'spo' &&
            this.config.get('componentType') === 'extension' &&
            this.config.get('extensionType') === 'SearchQueryModifier';
    }
}
exports.SearchQueryModifierGenerator = SearchQueryModifierGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = SearchQueryModifierGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
