// @microsoft/generator-sharepoint/lib/generators/app/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const colors = require("colors");
const path = require("path");
const yosay = require("yosay");
const spsay_1 = require("../../common/spsay");
const BaseGenerator_1 = require("../../common/BaseGenerator");
const Utils = require("../../common/utilities");
const YeomanConfiguration_1 = require("../../common/YeomanConfiguration");
const Solution = require("../solution");
const Component = require("../component");
const OnPrem19ApplicationCustomizer = require("../onPrem19ApplicationCustomizer");
const OnPrem19CommandSet = require("../onPrem19CommandSet");
const OnPrem19Component = require("../onPrem19Component");
const OnPrem19Extension = require("../onPrem19Extension");
const OnPrem19FieldCustomizer = require("../onPrem19FieldCustomizer");
const OnPrem19Webpart = require("../onPrem19Webpart");
const OnPremComponent = require("../onPremComponent");
const OnPremWebpart = require("../onPremWebpart");
const Webpart = require("../webpart");
const Extension = require("../extension");
const Library = require("../library");
const FieldCustomizer = require("../fieldCustomizer");
const CommandSet = require("../commandSet");
const SearchQueryModifier = require("../searchQueryModifier");
const ApplicationCustomizer = require("../applicationCustomizer");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
/**
 * This class is the entry point when running `yo @microsoft/sharepoint`
 */
class AppGenerator extends BaseGenerator_1.BaseGenerator {
    /* tslint:disable-next-line:no-any */
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'SharePoint Client-side Solution Generator';
        this.allowEmptyPackageJson = true;
        this.desc('Generate a SharePoint client-side solution with webparts.');
        Component.defineOptions(this);
        OnPremComponent.defineOptions(this);
        Solution.defineOptions(this);
    }
    initializing() {
        this.log(yosay('Welcome to the ' + colors.blue('SharePoint Client-side Solution Generator')));
        this._explainArguments();
        this.context.creatingSolution = !YeomanConfiguration_1.YeomanConfiguration.libraryName;
        if (this.options.solutionName !== undefined) {
            this.config.set('solutionName', this.options.solutionName);
        }
        if (this.options.environment !== undefined) {
            this.config.set('environment', this.options.environment);
        }
        if (this.options.packageManager !== undefined) {
            this.config.set('packageManager', this.options.packageManager);
        }
        if (this.options.skipFeatureDeployment !== undefined) {
            this.config.set('skipFeatureDeployment', this.options.skipFeatureDeployment);
        }
        if (this.options.isDomainIsolated !== undefined) {
            this.config.set('isDomainIsolated', this.options.isDomainIsolated);
        }
        if (this.options.componentType !== undefined) {
            this.config.set('componentType', this.options.environment === 'onprem' ? 'webpart' : this.options.componentType);
        }
        else {
            this.config.set('componentType', undefined);
        }
        if (this.options.extensionType !== undefined) {
            this.config.set('extensionType', this.options.extensionType);
        }
        else {
            this.config.set('extensionType', undefined);
        }
        if (this.options.framework !== undefined) {
            this.config.set('framework', this.options.framework);
        }
        if (this.options.componentName !== undefined) {
            this.config.set('componentName', this.options.componentName);
        }
        else {
            this.config.set('componentName', undefined);
        }
        if (this.options.componentDescription !== undefined) {
            this.config.set('componentDescription', this.options.componentDescription);
        }
        else {
            this.config.set('componentDescription', undefined);
        }
        if (this.options.plusbeta !== undefined) {
            this.config.set('plusBeta', this.options.plusbeta);
        }
        this.config.set('isCreatingSolution', this.context.creatingSolution);
        const baseOption = {
            'skip-install': this.options['skip-install'],
            plusbeta: this.options.plusbeta
        };
        // Config destination root before composition.
        Solution.composeWith(this, baseOption);
        OnPremComponent.composeWith(this, baseOption);
        OnPremWebpart.composeWith(this, baseOption);
        Component.composeWith(this, baseOption);
        Webpart.composeWith(this, baseOption);
        Extension.composeWith(this, baseOption);
        Library.composeWith(this, baseOption);
        FieldCustomizer.composeWith(this, baseOption);
        CommandSet.composeWith(this, baseOption);
        SearchQueryModifier.composeWith(this, baseOption);
        ApplicationCustomizer.composeWith(this, baseOption);
        OnPrem19Component.composeWith(this, baseOption);
        OnPrem19Webpart.composeWith(this, baseOption);
        OnPrem19Extension.composeWith(this, baseOption);
        OnPrem19FieldCustomizer.composeWith(this, baseOption);
        OnPrem19CommandSet.composeWith(this, baseOption);
        OnPrem19ApplicationCustomizer.composeWith(this, baseOption);
    }
    prompting() {
        // Configure the solution root folder before other composition.
        if (this.context.creatingSolution) {
            // Create solution because not in a solution folder
            this.log(`Let's create a new SharePoint solution.`);
        }
    }
    configuring() { }
    writing() { }
    install() { }
    end() {
        const libraryName = YeomanConfiguration_1.YeomanConfiguration.libraryName;
        this.log(spsay_1.default(libraryName));
    }
    _explainArguments() {
        // If the CLI argument is specified, the corresponding prompting in sub-generators will be skipped.
        // The following calls do similar jobs as prompting. Place them here instead of initializing step.
        if (this.options.solutionName) {
            const solutionName = colors.cyan(this.options.solutionName);
            this.log(`A folder with solution name ${solutionName} will be created for you.`);
        }
        if (this.options.environment) {
            const environment = colors.cyan(this.options.environment);
            this.log(`A component for the ${environment} environment will be created for you.`);
        }
        if (this.options.packageManager) {
            const packageManager = colors.cyan(this.options.packageManager);
            this.log(`The ${packageManager} package manager will be used.`);
        }
        const componentType = this.options.componentType === 'webpart' ? 'web part' :
            this.options.componentType === 'extension' ? 'extension' :
                this.options.componentType === 'library' ? 'library' : 'component';
        const componentTypeUpper = componentType.charAt(0).toUpperCase() + componentType.slice(1);
        if (this.options.componentName) {
            const webpartName = colors.cyan(this.options.componentName);
            const frameworkWords = this.options.framework && this.options.framework !== 'none'
                ? ' with ' + colors.cyan(this.options.framework) + ' framework'
                : '';
            this.log(`${componentType === 'extension' ? 'An' : 'A'} ${componentType} with the name ${webpartName}` +
                ` will be added to your solution${frameworkWords}.`);
        }
        if (this.options.componentDescription) {
            const description = colors.cyan(this.options.componentDescription);
            this.log(`${componentTypeUpper} description is going to be ${description}.`);
        }
    }
}
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = AppGenerator;
module.exports.composeWith = composeWith;
